File: edgeloop/cache.py
"""KV cache management for edgeloop.

Abstracts cache behavior so the agent loop can make cache-aware decisions
without knowing which backend is in use. Each backend reports its cache
state; the CacheManager uses that to decide when to truncate, what
format to send, and logs cache efficiency metrics.

The key insight for local LLMs: prefill is the bottleneck. Every token
we avoid re-prefilling is a direct latency win. This module exists to
minimize prefill at every opportunity.
"""

import logging
from dataclasses import dataclass, field

logger = logging.getLogger(__name__)


@dataclass
class CacheStats:
    """Stats from a single completion request."""
    prompt_tokens: int = 0
    generated_tokens: int = 0
    prefill_ms: float = 0.0
    generation_ms: float = 0.0
    cache_hit_tokens: int = 0  # tokens that were in cache (not re-prefilled)

    @property
    def prefill_tps(self) -> float:
        """Prefill tokens per second."""
        if self.prefill_ms <= 0:
            return 0.0
        return self.prompt_tokens / (self.prefill_ms / 1000)

    @property
    def generation_tps(self) -> float:
        """Generation tokens per second."""
        if self.generation_ms <= 0:
            return 0.0
        return self.generated_tokens / (self.generation_ms / 1000)

    @property
    def cache_hit_ratio(self) -> float:
        """Fraction of prompt tokens served from cache."""
        if self.prompt_tokens <= 0:
            return 0.0
        return self.cache_hit_tokens / self.prompt_tokens


@dataclass
class CacheManager:
    """Tracks KV cache state and makes cache-aware decisions.

    Each agent gets a CacheManager. It tracks:
    - How many tokens are currently cached (prefix length)
    - Cache hit ratios over time
    - When truncation is needed to stay within context budget

    The CacheManager doesn't talk to backends directly — it receives
    stats after each completion and advises the agent on what to do.
    """

    max_context_tokens: int = 4096
    system_prompt_tokens: int = 0  # set once at agent init
    _history_tokens: int = 0
    _total_requests: int = 0
    _total_cache_hits: int = 0
    _total_prompt_tokens: int = 0
    _last_stats: CacheStats | None = None
    _per_request_stats: list[CacheStats] = field(default_factory=list)

    def record(self, stats: CacheStats):
        """Record stats from a completion request."""
        self._total_requests += 1
        self._total_cache_hits += stats.cache_hit_tokens
        self._total_prompt_tokens += stats.prompt_tokens
        self._last_stats = stats
        self._per_request_stats.append(stats)

        logger.debug(
            "Cache: prefill=%d tok (%.0fms), gen=%d tok, cache_hit=%.0f%%",
            stats.prompt_tokens, stats.prefill_ms,
            stats.generated_tokens, stats.cache_hit_ratio * 100,
        )

    def update_history_tokens(self, token_count: int):
        """Update the estimated token count of the current history."""
        self._history_tokens = token_count

    @property
    def total_tokens(self) -> int:
        """Estimated total tokens in current prompt (system + history)."""
        return self.system_prompt_tokens + self._history_tokens

    @property
    def remaining_tokens(self) -> int:
        """Tokens available before hitting context limit."""
        return max(0, self.max_context_tokens - self.total_tokens)

    @property
    def needs_truncation(self) -> bool:
        """True if we're over 80% of context budget — truncate soon."""
        return self.total_tokens > self.max_context_tokens * 0.8

    @property
    def overall_cache_hit_ratio(self) -> float:
        """Lifetime cache hit ratio across all requests."""
        if self._total_prompt_tokens <= 0:
            return 0.0
        return self._total_cache_hits / self._total_prompt_tokens

    def truncation_target(self) -> int:
        """How many history messages to keep after truncation.

        Strategy: keep system prompt + most recent messages that fit
        in 60% of context budget. This leaves room for the next
        tool call cycle without another truncation.
        """
        target_tokens = int(self.max_context_tokens * 0.6) - self.system_prompt_tokens
        return max(target_tokens, 0)

    def summary(self) -> dict:
        """Return a summary of cache performance."""
        return {
            "total_requests": self._total_requests,
            "total_prompt_tokens": self._total_prompt_tokens,
            "total_cache_hits": self._total_cache_hits,
            "cache_hit_ratio": self.overall_cache_hit_ratio,
            "current_context_tokens": self.total_tokens,
            "max_context_tokens": self.max_context_tokens,
            "last_prefill_ms": self._last_stats.prefill_ms if self._last_stats else 0,
        }

    def log_summary(self):
        """Log a human-readable cache performance summary."""
        s = self.summary()
        logger.info(
            "Cache summary: %d requests, %.0f%% cache hit ratio, "
            "%d/%d context tokens, last prefill %.0fms",
            s["total_requests"],
            s["cache_hit_ratio"] * 100,
            s["current_context_tokens"],
            s["max_context_tokens"],
            s["last_prefill_ms"],
        )

File: src/cache.rs
use serde::Serialize;
use tracing::{debug, info};

#[derive(Debug, Clone, Default)]
pub struct CacheStats {
    pub prompt_tokens: usize,
    pub generated_tokens: usize,
    pub prefill_ms: f64,
    pub generation_ms: f64,
    pub cache_hit_tokens: usize,
}

impl CacheStats {
    pub fn cache_hit_ratio(&self) -> f64 {
        if self.prompt_tokens == 0 { return 0.0; }
        self.cache_hit_tokens as f64 / self.prompt_tokens as f64
    }
}

#[derive(Debug, Serialize)]
pub struct CacheSummary {
    pub total_requests: usize,
    pub total_prompt_tokens: usize,
    pub cache_hit_ratio: f64,
    pub current_context_tokens: usize,
    pub max_context_tokens: usize,
    pub last_prefill_ms: f64,
}

#[derive(Debug)]
pub struct CacheManager {
    pub max_context_tokens: usize,
    pub system_prompt_tokens: usize,
    pub truncation_threshold: f64,
    history_tokens: usize,
    total_requests: usize,
    total_cache_hits: usize,
    total_prompt_tokens: usize,
    last_stats: Option<CacheStats>,
}

impl CacheManager {
    pub fn new(max_context: usize, truncation_threshold: f64) -> Self {
        Self {
            max_context_tokens: max_context,
            system_prompt_tokens: 0,
            truncation_threshold,
            history_tokens: 0,
            total_requests: 0,
            total_cache_hits: 0,
            total_prompt_tokens: 0,
            last_stats: None,
        }
    }

    pub fn record(&mut self, stats: CacheStats) {
        self.total_requests += 1;
        self.total_cache_hits += stats.cache_hit_tokens;
        self.total_prompt_tokens += stats.prompt_tokens;
        debug!(
            "Cache: prefill={} tok ({:.0}ms), gen={} tok, hit={:.0}%",
            stats.prompt_tokens, stats.prefill_ms, stats.generated_tokens,
            stats.cache_hit_ratio() * 100.0,
        );
        self.last_stats = Some(stats);
    }

    pub fn update_history_tokens(&mut self, count: usize) {
        self.history_tokens = count;
    }

    pub fn total_tokens(&self) -> usize {
        self.system_prompt_tokens + self.history_tokens
    }

    pub fn remaining_tokens(&self) -> usize {
        self.max_context_tokens.saturating_sub(self.total_tokens())
    }

    pub fn needs_truncation(&self) -> bool {
        self.total_tokens() as f64 > self.max_context_tokens as f64 * self.truncation_threshold
    }

    pub fn truncation_target(&self) -> usize {
        let target = (self.max_context_tokens as f64 * 0.6) as usize;
        target.saturating_sub(self.system_prompt_tokens)
    }

    pub fn overall_cache_hit_ratio(&self) -> f64 {
        if self.total_prompt_tokens == 0 { return 0.0; }
        self.total_cache_hits as f64 / self.total_prompt_tokens as f64
    }

    pub fn summary(&self) -> CacheSummary {
        CacheSummary {
            total_requests: self.total_requests,
            total_prompt_tokens: self.total_prompt_tokens,
            cache_hit_ratio: self.overall_cache_hit_ratio(),
            current_context_tokens: self.total_tokens(),
            max_context_tokens: self.max_context_tokens,
            last_prefill_ms: self.last_stats.as_ref().map(|s| s.prefill_ms).unwrap_or(0.0),
        }
    }

    pub fn log_summary(&self) {
        let s = self.summary();
        info!(
            "Cache: {} requests, {:.0}% hit, {}/{} tokens, last prefill {:.0}ms",
            s.total_requests, s.cache_hit_ratio * 100.0,
            s.current_context_tokens, s.max_context_tokens, s.last_prefill_ms,
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cache_basics() {
        let mut cm = CacheManager::new(4096, 0.8);
        cm.system_prompt_tokens = 100;
        cm.update_history_tokens(200);
        assert_eq!(cm.total_tokens(), 300);
        assert_eq!(cm.remaining_tokens(), 3796);
        assert!(!cm.needs_truncation());
    }

    #[test]
    fn test_truncation_needed() {
        let mut cm = CacheManager::new(1000, 0.8);
        cm.system_prompt_tokens = 100;
        cm.update_history_tokens(750);
        assert!(cm.needs_truncation());
    }

    #[test]
    fn test_record_stats() {
        let mut cm = CacheManager::new(4096, 0.8);
        cm.record(CacheStats {
            prompt_tokens: 100,
            generated_tokens: 20,
            prefill_ms: 15.0,
            generation_ms: 50.0,
            cache_hit_tokens: 80,
        });
        assert_eq!(cm.total_requests, 1);
        assert_eq!(cm.overall_cache_hit_ratio(), 0.8);
    }
}
